astropy/units/quantity.py

```diff
@@ -384,6 +386,11 @@ def __new__(cls, value, unit=None, dtype=None, copy=True, order=None,
 
         # by default, cast any integer, boolean, etc., to float
         if dtype is None and value.dtype.kind in 'iuO':
+            if not copy:
+                warnings.warn(f'Casting {value.dtype} to float, which will '
+                              'implicitly make a copy of the data. '
+                              'This can be prevented by explicitly setting the '
+                              'dtype argument.')
```

Comment:
Thoughts on raising a warning here? I think it's good if a user has explicitly specified `copy=False` to warn them if a copy is being made.

Reply:
If we go that route, we'd need some way to avoid the warning for things like `np.arange(10) << u.m`...

Reply:
It should be fine to do casting in `__lshift__` before handing to `__class__(self, other, copy=False, subok=True)`.

Do you think having the warning is useful? If so I can go ahead and modify `__lshift__` (and I think there's a couple of other places that need checking too)

Reply:
I'm not keen in duplicating the `Quantity` casting behaviour inside a method on `UnitBase` - it is non-trivial (see further down). 

I think if we go this route, I'd prefer a different `copy` option (perhaps `None` for raise instead of converting). See https://github.com/astropy/astropy/issues/10079#issuecomment-605454494

Maybe it is best to keep this separate, though: I think a change in the docstring to clarify the current behaviour is good regardless!